OpenRA.Mods.Cnc/Projectiles/TeslaZap.cs

```diff
@@ -32,46 +32,48 @@ public class TeslaZapInfo : IProjectileInfo
 
 		public readonly int Duration = 2;
 
+		public readonly int DamageDuration = 1;
+
+		public readonly bool TrackTarget = true;
+
 		public IProjectile Create(ProjectileArgs args) { return new TeslaZap(this, args); }
 	}
 
-	public class TeslaZap : IProjectile
+	public class TeslaZap : IProjectile, ISync
 	{
 		readonly ProjectileArgs args;
 		readonly TeslaZapInfo info;
 		TeslaZapRenderable zap;
-		int timeUntilRemove; // # of frames
-		bool doneDamage = false;
-		bool initialized = false;
+		int ticksUntilRemove;
+		int damageDuration;
+		[Sync] WPos target;
 
 		public TeslaZap(TeslaZapInfo info, ProjectileArgs args)
 		{
 			this.args = args;
 			this.info = info;
-			this.timeUntilRemove = info.Duration;
+			ticksUntilRemove = info.Duration;
+			damageDuration = info.DamageDuration > info.Duration ? info.Duration : info.DamageDuration;
+			target = args.PassiveTarget;
 		}
 
 		public void Tick(World world)
 		{
-			if (timeUntilRemove-- <= 0)
+			if (ticksUntilRemove-- <= 0)
 				world.AddFrameEndTask(w => w.Remove(this));
 
-			if (!doneDamage)
-			{
-				var pos = args.GuidedTarget.IsValidFor(args.SourceActor) ? args.GuidedTarget.CenterPosition : args.PassiveTarget;
-				args.Weapon.Impact(Target.FromPos(pos), args.SourceActor, args.DamageModifiers);
-				doneDamage = true;
-			}
+			// Zap tracks target
+			if (info.TrackTarget && args.GuidedTarget.IsValidFor(args.SourceActor))
+				target = args.GuidedTarget.Positions.PositionClosestTo(args.Source);
```

Comment:
`TrackTarget` cannot change during the game.
About `IsValidFor`: Since the target might have moved since the projectile launch, the zap would jump back to the original target position if we used `args.PassiveTarget` instead of last `target` position. This would look glitchy, especially for fast targets. Using the last target position before it became invalid is the safest approach here, in my opinion.